### DISCUSSION TITLE
Feature Request: Support Promise for custom function callback

### DISCUSSION THREAD

**Discussion by ymor** [FontoXML/fontoxpath#657] [Ideas]
Hi,
I assume this will be non trivial.  But we are looking at adding digest support via a custom function using the Web Crypto API https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest 

However the methods are all promise based, which currently doesn't work with the custom function register callback.
So I was wondering what sort of effort would be involved to support this?

Thanks.

**Comment by DrRataplan** [comment]
Hey,

Huge. Those promises will have to bubble up all the way to the top level,
so the whole code path from that function call to the `evaluateXPath`
function needs to handle async code. We used to have something for this
(it's still remnant in the `evaluatexpathtoasynciterator` code (
https://github.com/FontoXML/fontoxpath/blob/master/src/evaluateXPathToAsyncIterator.ts),
but we ended up not using it. So that whole path was (from Fonto) only used
in unit tests. Don't forget, all the places where that async
`evaluateXPathToAsyncIterator` is used also needs to be asynchronous.
Something about releasing Zalgo.

So, instead, we went for this: a promise is nothing more than a function
that returns a 'loading' placeholder at first, and a 'resolved value'
later. Now you only need a signal that the load is completed and your
function is invalidated. In Fonto we have a stable dependency tracking /
indexing framework we can utilise for that.

But you don't have a dependency framework in place, and Fonto's is not open
sourced. You will have to do something else.

I'd approach it like this:

   1. A single event source / pub-sub / notifier / however you want to call
   it that can be called to signal that your XPath changed.
   2. In a custom function, see if there is an earlier promise already
   resolved. If so, just return the resolved value, if not, start a request
   and return a 'loading' placeholder. We used a pattern like `map{"status":
   "resolved" | "in-flight", "value": "the resolved thing"}`
   3. After the request is done, notify that to your pub/sub whatever you
   made in step #1
   4. Repeat until there are no more requests in flight (or in the case of
   an application, forever: stuff can change, other things can invalidate your
   event source, etcetera)

Are you building an application that updates itself every so often or
something that just runs once?

Kindest regards,

Martin

**Comment by ymor** [comment]
Thanks for the prompt and detailed response. I had assumed as much :) 

Its essentially an XForms based application so potentially plenty of updates. I will look into your suggestion further.
Thanks again.

**Comment by DrRataplan** [comment]
XForms, cool! Check Fore (https://github.com/jinntec/Fore). It's a project
I am also working with. We did not go into promises yet, but it's always
fun to look at other solutions that take roughly the same path.

**Reply by ymor** [reply]
Nice! Also good to see someone else use the word `xformish` 😁
